File: app/src/main/java/org/cosmic/ide/ui/logger/Logger.java
package org.cosmic.ide.ui.logger;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

public class Logger {

    private LogAdapter adapter;
    private List<Log> data = new ArrayList<>();

    private RecyclerView mRecyclerView;

    public void attach(RecyclerView view) {
        mRecyclerView = view;
        init();
    }

    private void init() {
        adapter = new LogAdapter(data);
        var layoutManager = new LinearLayoutManager(mRecyclerView.getContext());
        layoutManager.setStackFromEnd(true);
        mRecyclerView.setLayoutManager(layoutManager);
        mRecyclerView.setAdapter(adapter);
    }

    public void message(String message) {
        message(null, message);
    }

    public void message(String tag, String message) {
        mRecyclerView.post(
                () -> {
                    data.add(new Log(tag, message));
                    adapter.notifyItemInserted(data.size());
                    mRecyclerView.smoothScrollToPosition(data.size() - 1);
                });
    }
}

File: app/src/main/java/org/cosmic/ide/ui/logger/Logger.kt
package org.cosmic.ide.ui.logger

import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

class Logger {
    private var adapter: LogAdapter? = null
    private val data: MutableList<Log> = ArrayList()
    private var mRecyclerView: RecyclerView? = null
    fun attach(view: RecyclerView?) {
        mRecyclerView = view
        init()
    }

    private fun init() {
        adapter = LogAdapter(data)
        val layoutManager = LinearLayoutManager(
            mRecyclerView!!.context
        )
        layoutManager.stackFromEnd = true
        mRecyclerView!!.layoutManager = layoutManager
        mRecyclerView!!.adapter = adapter
    }

    fun message(message: String?) {
        mRecyclerView!!.post {
            data.add(Log(null, message!!))
            adapter!!.notifyItemInserted(data.size)
            mRecyclerView!!.smoothScrollToPosition(data.size - 1)
        }
    }
}